Add cascade delete and update for services

calvinbrewer opened the pull request.

calvinbrewer committed:
Add cascade delete and update for services

calvinbrewer committed:
Add cascade delete function

dylnslck commented on src/services/database/cascadeDelete.js:
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable no-param-reassign */
+import schemas from '../../schemas';
+import relationshipFinder from '../utils/relationshipFinder';
```
This way of naming utilities is a little unconventional. It should really be called something like `getRelationships`, because then it's clear that it's going to return something.

dylnslck commented on src/services/database/cascadeDelete.js:
```diff
@@ -0,0 +1,62 @@
+
+  for (const link of relationships) {
+    if (link.table === table) {
+      return;
```
If you're going to `return` immediately, don't place it above an `else if`. Short circuit it immediately: 

```
if (link.table === table) return;
```

That way the indentations don't get too insane, and curly braces are hard on the eyes.

dylnslck commented on src/services/database/cascadeDelete.js:
```diff
@@ -0,0 +1,62 @@
+
+  const relationships = relationshipFinder(table, schema);
+
+  for (const link of relationships) {
```
In my opinion, `link` should actually be named `relationship`.

calvinbrewer replied:
I had `relationship` but `link` makes the code more readable

dylnslck commented on src/services/database/cascadeDelete.js:
```diff
@@ -0,0 +1,62 @@
+
+const promise = [];
+
+const traverseDelete = (record, table, schema) => {
```
What happened to using `previousRelationship` as the third argument. Is that not valid for some reason?

calvinbrewer replied:
`table` is the previous relationship

dylnslck commented on src/services/database/cascadeDelete.js:
```diff
@@ -0,0 +1,62 @@
+import relationshipFinder from '../utils/relationshipFinder';
+import patch from '../utils/patchRelationship';
+
+const promise = [];
```
This name should be plural and maybe more explicit. A promise of what? Maybe something like `deletionPromises`? Also, I'm not sure that should be a global variable.

jebukowski committed:
Add services util and install lodash

jebukowski committed:
Create test for and fix `getRelationships`

calvinbrewer committed:
Refactor cascadeDelete

calvinbrewer committed:
Merge with branch

dylnslck commented on src/services/utils/getRelationships.js:
```diff
@@ -0,0 +1,38 @@
+import _ from 'lodash';
```
`import { forEach } from lodash`

dylnslck commented on test/services/utils/getRelationships.test.js:
```diff
@@ -0,0 +1,40 @@
+      inverse: 'belongsTo',
+    },
+  ];
+
```
```
embeddedResult = {
  brands: {
    table: 'brand',
    original: 'belongsTo',
    inverse: 'hasMany',
  },
},
```

calvinbrewer committed:
Add test script for cascade tests

jebukowski committed:
Modify `getRelationships` and update its test

The returned entity was changed to an object so that relationship fields would be easier to access.

jebukowski committed:
Update semantics (`name` to `field`) and JSDoc

calvinbrewer committed:
Update cascade archive

calvinbrewer committed:
Add inverse tag to schemas and fixtures

jebukowski committed:
Update `getRelationships` to include inverse field

calvinbrewer committed:
Refactor cascade archive

Add cascade queries utils

calvinbrewer committed:
Add JSDoc to cascade functions

calvinbrewer committed:
Fix JSDoc tabs

Add more descriptive descriptions

calvinbrewer committed:
Add cascadePost and start refactoring all tests

jebukowski committed:
Create initial version of `getArchiveObject`

jebukowski committed:
Fix merge

jebukowski committed:
Remove test/services/utils/getArchiveObject.test.js

jebukowski committed:
Correct `getArchiveObject` and modularize helpers

Put the `getArchiveObject` helpers in their own files to reduce clutter.
Generally tidied up the code and created tests for all relevant files.

calvinbrewer committed:
Add serialize response and test

jebukowski committed:
Update `sanitizeRequest` to handle patch request

jebukowski committed:
Update `sanitizeRequest` and its test

Resolves: #1

calvinbrewer committed:
Add cascade update

calvinbrewer committed:
Merge branch 'services-cascade' of github.com:endfire/webcom-server into services-cascade

jebukowski committed:
Update `schemas`

The table `data` was changed to `submission` for clarity. Users will
actually be submitting a form which will store data, not the other
way around.

Resolves: #30

calvinbrewer committed:
Implement redink to replace database services

This closes #1 and closes #29.

calvinbrewer committed:
Merge branch 'services-cascade' of github.com:endfire/webcom-server into services-cascade
